adala/skills/collection/prompt_improvement.py

```diff
@@ -0,0 +1,149 @@
+
+
+# NOTE: these response models are converging with the LSE ResultHandler, slowly pushing typing deeper into the lib with the end goal of combining them
+
+
+class PromptImprovementSkillResponseModel(BaseModel):
+
+    # hidden variable, used for validation
+    _input_variables: List[str]
+    reasoning: str
+    improved_user_prompt: str
+    # NOTE: not exposed in LSE yet, so default is always used. Should improve this as well when we expose it.
+    # improved_system_prompt: str
+
+    model_config = ConfigDict(
+        # omit other fields
+        extra="ignore",
+        # guard against name collisions with other fields
+        populate_by_name=False,
+    )
+
+    @field_validator("improved_user_prompt", mode="after")
+    def validate_used_variables(cls, value: str) -> str:
+
+        templates = parse_template(value, include_texts=False)
+        if not templates:
+            raise ValueError("At least one input variable must be used in the prompt")
+
+        input_vars_used = [t["text"] for t in templates]
+        if extra_vars_used := set(input_vars_used) - set(cls._input_variables):
+            raise ValueError(
+                f"Invalid variable used in prompt: {extra_vars_used}. Valid variables are: {cls._input_variables}"
+            )
+
+        return value
+
+
+class ErrorResponseModel(BaseModel):
+    message: str = Field(..., alias="_adala_message")
+    details: str = Field(..., alias="_adala_details")
+
+    model_config = ConfigDict(
+        # omit other fields
+        extra="ignore",
+        # guard against name collisions with other fields
+        populate_by_name=False,
+    )
+
+
+class ImprovedPromptResponse(BaseModel):
+
+    output: Union[PromptImprovementSkillResponseModel, ErrorResponseModel]
+
+    prompt_tokens: int = Field(alias="_prompt_tokens")
+    completion_tokens: int = Field(alias="_completion_tokens")
+
+    # these can fail to calculate
+    prompt_cost_usd: Optional[float] = Field(alias="_prompt_cost_usd")
+    completion_cost_usd: Optional[float] = Field(alias="_completion_cost_usd")
+    total_cost_usd: Optional[float] = Field(alias="_total_cost_usd")
+
+
+def get_prompt_improvement_inputs(
+    student_skill: Skill, input_variables: List[str], student_model: str
+) -> Dict[str, Any]:
+    return {
+        "model": student_model,
+        "task_name": student_skill.name,
+        "task_description": student_skill.description,
+        "input_variables": input_variables,
+        "current_system_prompt": student_skill.instructions,
+        "current_user_prompt": student_skill.input_template,
+        "response_json_schema": student_skill.response_model.model_json_schema(),
+    }
+
+
+def get_prompt_improvement_skill(input_variables: List[str]) -> TextGenerationSkill:
+
+    # setting this dynamically - used to validate the improved prompt
+    PromptImprovementSkillResponseModel._input_variables = input_variables
+
+    prompt_improvement_skill = TextGenerationSkill(
```

Comment:
We have `AnalysisSkill` that effectively transforms table into a single generated string